elastic-agent standalone on openshift documentation (#1498)
elastic/observability-docs: docs/en/ingest-management/elastic-agent/running-on-kubernetes-standalone.asciidoc

elastic-agent-hs4pb             1/1     Running   0          81m
------------------------------------------------

[discrete]
== Red Hat OpenShift configuration

If you are using Red Hat OpenShift, you need to specify additional settings in
the manifest file and enable the container to run as privileged.

. In the manifest file, modify the `agent-node-datastreams` ConfigMap and adjust inputs:
+
--
* `kubernetes-cluster-metrics` input:

> ** If `https` is used to access `kube-state-metrics`, then to all `kubernetes.state_*` datasets should be added:

Reviewer: ```suggestion
** If `https` is used to access `kube-state-metrics`, add the following settings to all `kubernetes.state_*` datasets:
```

+
[source,yaml]
------------------------------------------------
  bearer_token_file: /var/run/secrets/kubernetes.io/serviceaccount/token
  ssl.certificate_authorities:
    - /var/run/secrets/kubernetes.io/serviceaccount/service-ca.crt
------------------------------------------------
* `kubernetes-node-metrics` input:

> ** `kubernetes.controllermanager` datastream condition should be change to:

Reviewer: ```suggestion
** Change the `kubernetes.controllermanager` datastream condition to:
```

+
[source,yaml]
------------------------------------------------
condition: ${kubernetes.labels.app} == 'kube-controller-manager'
------------------------------------------------

> ** `kubernetes.scheduler` datastream condition should be changed to:

Reviewer: ```suggestion
** Change the `kubernetes.scheduler` datastream condition to:
```

+
[source,yaml]
------------------------------------------------
condition: ${kubernetes.labels.app} == 'openshift-kube-scheduler'
------------------------------------------------

> ** `kubernetes.proxy` datastream configutation should look like:

Reviewer: ```suggestion
** The `kubernetes.proxy` datastream configuration should look like:
```

+
[source,yaml]
------------------------------------------------
- data_stream:
    dataset: kubernetes.proxy
    type: metrics
  metricsets:
    - proxy
  hosts:
    - 'localhost:29101'
  period: 10s
------------------------------------------------

> ** to all datastreams that connnect to `https://${env.NODE_NAME}:10250` should be added:

Reviewer: ```suggestion
** Add the following settings to all datastreams that connect to `https://${env.NODE_NAME}:10250`:
```

+
[source,yaml]
------------------------------------------------
  bearer_token_file: /var/run/secrets/kubernetes.io/serviceaccount/token
  ssl.certificate_authorities:
    - /path/to/ca-bundle.crt
------------------------------------------------
NOTE: `ca-bundle.crt` can be any CA bundle that contains the issuer of the certificate used in the Kubelet API.
According to each specific installation of Openshift this can be found either in `secrets` or in `configmaps`.
In some installations it can be available as part of the service account secret, in
`/var/run/secrets/kubernetes.io/serviceaccount/service-ca.crt`.
In case of using https://github.com/openshift/installer/blob/master/docs/user/gcp/install.md[Openshift installer]

> for GCP then the following `configmap` can be mounted in Metricbeat Pod and use `ca-bundle.crt`

Reviewer 2: Did you mean elastic-agent pod?

Author: thank you for pointing it out! done - https://github.com/elastic/observability-docs/pull/1498/commits/eca949819181efb5fff36f948943d03b5f389b87

> In case of using https://github.com/openshift/installer/blob/master/docs/user/gcp/install.md[Openshift installer]
> for GCP then the following `configmap` can be mounted in elastic-agent pod and use `ca-bundle.crt`
> in `ssl.certificate_authorities`:

Reviewer: ```suggestion
When using the https://github.com/openshift/installer/blob/master/docs/user/gcp/install.md[Openshift installer]
for GCP, mount the following `configmap` in the elastic-agent pod and use `ca-bundle.crt`
in `ssl.certificate_authorities`:
```

+
[source,shell]
-----
Name:         kubelet-serving-ca
Namespace:    openshift-kube-apiserver
Labels:       <none>
Annotations:  <none>

Data
====
ca-bundle.crt:
-----
--
. Grant the `elastic-agent` service account access to the privileged SCC:
+
[source,shell]
-----
oc adm policy add-scc-to-user privileged system:serviceaccount:kube-system:elastic-agent
-----
+
This command enables the container to be privileged as an administrator for
OpenShift.

. If namespace, where elastic-agent is running, has `"openshift.io/node-selector"` annotation set, elastic-agent

> might get running not on all nodes. In this case consider overriding node selector for the namespace to allow

Reviewer 2: NIT: might not run on all nodes

> . If namespace, where elastic-agent is running, has `"openshift.io/node-selector"` annotation set, elastic-agent
> might not run on all nodes. In this case consider overriding node selector for the namespace to allow scheduling
> on any node:

Reviewer: ```suggestion
. If the namespace where elastic-agent is running has the `"openshift.io/node-selector"` annotation set, elastic-agent
might not run on all nodes. In this case consider overriding the node selector for the namespace to allow scheduling
on any node:
```

elastic-agent-hs4pb             1/1     Running   0          81m
------------------------------------------------

[discrete]
== Red Hat OpenShift configuration

If you are using Red Hat OpenShift, you need to specify additional settings in
the manifest file and enable the container to run as privileged.

. In the manifest file, modify the `agent-node-datastreams` ConfigMap and adjust inputs:
`kubernetes-cluster-metrics` input:
- If used https to access kube-state-metrics, to all `kubernetes.state_*` dataset should be added:
+
[source,yaml]
------------------------------------------------
  bearer_token_file: /var/run/secrets/kubernetes.io/serviceaccount/token
  ssl.certificate_authorities:
    - /var/run/secrets/kubernetes.io/serviceaccount/service-ca.crt
------------------------------------------------

> `kubernetes-node-metrics` input:

Reviewer: I won't have time to review this until next week, but I wanted to mention that the formatting in this section doesn't look quite right. There's some useful documentation about complex list items here that might help you: https://docs.asciidoctor.org/asciidoc/latest/lists/continuation/

I think their examples are a little confusing, but I personally find it easier to use  open blocks to create a complex list. For example:

```
. I'm a numbered list item.
+
--
I'm a bunch of content nested under the list item.

I can be multiple paragraphs, code examples, lists, and so on.

I cannot contain another open block.
--
```

Author: Thank you for the hint! I've updated formatting

elastic-agent-hs4pb             1/1     Running   0          81m
------------------------------------------------

[discrete]
== Red Hat OpenShift configuration

If you are using Red Hat OpenShift, you need to specify additional settings in
the manifest file and enable the container to run as privileged.

. In the manifest file, modify the `agent-node-datastreams` ConfigMap and adjust inputs:
+
--
* `kubernetes-cluster-metrics` input:

> ** If used https to access kube-state-metrics, to all `kubernetes.state_*` dataset should be added:

Reviewer 3: ```suggestion
** If `https` is used to access `kube-state-metrics`, then to all `kubernetes.state_*` datasets should be added:
```

  ssl.certificate_authorities:
    - /path/to/ca-bundle.crt
------------------------------------------------
NOTE: `ca-bundle.crt` can be any CA bundle that contains the issuer of the certificate used in the Kubelet API.
According to each specific installation of Openshift this can be found either in `secrets` or in `configmaps`.
In some installations it can be available as part of the service account secret, in
`/var/run/secrets/kubernetes.io/serviceaccount/service-ca.crt`.
In case of using https://github.com/openshift/installer/blob/master/docs/user/gcp/install.md[Openshift installer]
for GCP then the following `configmap` can be mounted in Metricbeat Pod and use `ca-bundle.crt`
in `ssl.certificate_authorities`:
+
[source,shell]
-----
Name:         kubelet-serving-ca
Namespace:    openshift-kube-apiserver
Labels:       <none>
Annotations:  <none>

Data
====
ca-bundle.crt:
-----
--
. Grant the `elastic-agent` service account access to the privileged SCC:
+
[source,shell]
-----
oc adm policy add-scc-to-user privileged system:serviceaccount:kube-system:elastic-agent
-----
+
This command enables the container to be privileged as an administrator for
OpenShift.

> . If namespace, where elastic-agent is running, has `"openshift.io/node-selector"` annotation set, override node

Reviewer 3: Isn't this change too generic? This might affect other workloads too in the cluster no? Any alternatives here?

Author: I took this part from the metricbeat documentation - https://www.elastic.co/guide/en/beats/metricbeat/current/running-on-kubernetes.html#_red_hat_openshift_configuration

actually this can be just dropped, because on Openshift v4 in default namespace for elastic-agent - `kube-system` this annotation is not set (openshift on gcp setup), so elastic-agent was deployed on both worker and master nodes without any changes. I kept this just as a possible explanation why elastic-agent might be not deployed on all nodes (if running not in standard namespace or there are some differences on other environments)

wdyt about:
```suggestion
. If namespace, where elastic-agent is running, has `"openshift.io/node-selector"` annotation set, elastic-agent might get running not on all nodes. In this case consider overriding node selector for the namespace to allow scheduling on any node:
```

Reviewer 3: Thanks for elaborating, didn't remember we had this in Beats too. I think your suggestion is safer, so good to go!

Author: done https://github.com/elastic/observability-docs/pull/1498/commits/1a048d1f1acda4f6730c9b509e3ec5a3c4705b2a
